Grbl_Esp32/src/Pin.cpp

```diff
@@ -0,0 +1,173 @@
+#include "Pin.h"
+
+// Pins:
+#include "Pins/PinOptionsParser.h"
+#include "Pins/VoidPinDetail.h"
+#include "Pins/GPIOPinDetail.h"
+#include "Pins/I2SOPinDetail.h"
+
+#if defined PIN_DEBUG && defined ESP32
+#    include "Pins/DebugPinDetail.h"
+#endif
+
+#ifdef ESP32
+#    include "Grbl.h"  // grbl_sendf
+#endif
+
+bool Pin::parse(String str, Pins::PinDetail*& pinImplementation) {
+    // Initialize pinImplementation first! Callers might want to delete it, and we don't want a random pointer.
+    pinImplementation = nullptr;
+
+    // Parse the definition: [GPIO].[pinNumber]:[attributes]
+
+    // Skip whitespaces at the start
+    auto nameStart = str.begin();
+    for (; nameStart != str.end() && ::isspace(*nameStart); ++nameStart) {}
+
+    if (nameStart == str.end()) {
+        // Re-use undefined pins happens in 'create':
+        pinImplementation = new Pins::VoidPinDetail();
+        return true;
+    }
+
+    auto idx = nameStart;
+    for (; idx != str.end() && *idx != '.' && *idx != ':'; ++idx) {
+        *idx = char(::tolower(*idx));
+    }
+
+    String prefix = str.substring(0, int(idx - str.begin()));
+
+    if (idx != str.end()) {  // skip '.'
+        ++idx;
+    }
+
+    int pinNumber = 0;
+    if (prefix != "") {
+        if (idx == str.end()) {
+            // Incorrect pin definition.
+            return false;
+        }
+
+        for (; idx != str.end() && *idx >= '0' && *idx <= '9'; ++idx) {
+            pinNumber = pinNumber * 10 + int(*idx - '0');
+        }
+        if (pinNumber < 0 || pinNumber > 253) {
+            // Pin number has to be between [0,253].
+            return false;
+        }
+    }
+
+    String options;
+    if (idx != str.end()) {
```

Comment:
if (idx != str.end() && *idx > ' ') {

this fixed the issue with pin settings from a file having a <CR> at the end. Probably not the right fix, but at least pin settings from files are working.